File: ios/Classes/ObservableSourceHolder.swift
import RxSwift

protocol ObservableSourceHolder {
    var type: StreamType {get set}
    var sourceGenerator: (Any?) throws -> Any {get set}
    var errorHandler: ((Error) -> Any?)? {get set}
    
    func getSourceAsObservable(
        _ args: Any?
        ) throws -> Observable<Any>
    
    func getSourceAsSingle(
        _ args: Any?
        ) throws -> PrimitiveSequence<SingleTrait, Any>
    
    func getSourceAsCompletable(
        _ args: Any?
        ) throws -> PrimitiveSequence<CompletableTrait, Never>
}

class ObservableSourceHolderImpl: ObservableSourceHolder {
    var type: StreamType
    var sourceGenerator: (Any?) throws -> Any
    var errorHandler: ((Error) -> Any?)?
    
    init(
        observable: @escaping (Any?) throws -> Any,
        _ errorHandler: ((Error) -> Any?)? = nil
        ) {
        self.type = StreamType.OBSERVABLE
        self.sourceGenerator = observable
        self.errorHandler = errorHandler
    }
    
    init(
        single: @escaping (Any?) throws -> Any,
        _ errorHandler: ((Error) -> Any?)? = nil
        ) {
        self.type = StreamType.SINGLE
        self.sourceGenerator = single
        self.errorHandler = errorHandler
    }
    
    init(
        completable: @escaping (Any?) throws -> Any,
        _ errorHandler: ((Error) -> Any?)? = nil
        ) {
        self.type = StreamType.COMPLETABLE
        self.sourceGenerator = completable
        self.errorHandler = errorHandler
    }
    
    func getSourceAsObservable(
        _ args: Any?
        ) throws -> Observable<Any> {
        return try sourceGenerator(args) as! Observable<Any>
    }
    
    func getSourceAsSingle(
        _ args: Any?
        ) throws -> PrimitiveSequence<SingleTrait, Any> {
        return try sourceGenerator(args) as! PrimitiveSequence<SingleTrait, Any>
    }
    
    func getSourceAsCompletable(
        _ args: Any?
        ) throws -> PrimitiveSequence<CompletableTrait, Never> {
        return try sourceGenerator(args) as! PrimitiveSequence<CompletableTrait, Never>
    }
}

File: android/src/main/kotlin/tenko/rx_flutter_plugin/ObservableSourceHolder.kt
package tenko.rx_flutter_plugin

import io.reactivex.CompletableSource
import io.reactivex.ObservableSource
import io.reactivex.SingleSource

class ObservableSourceHolder<T> {
    internal lateinit var type: StreamType
    internal lateinit var sourceGenerator: (args: T?) -> Any
    internal var errorHandler: ((Throwable) -> Any?)? = null

    //Not in use atm.
//    class Builder<T> {
//        val holder = ObservableSourceHolder<T>()
//
//        fun build(): ObservableSourceHolder<T> {
//            return holder
//        }
//    }

    companion object {
        fun <T> observableHolder(observable: (arguments: T?) -> ObservableSource<*>, errorHandler: ((Throwable) -> Any?)? = null): ObservableSourceHolder<T> {
            val holder =  ObservableSourceHolder<T>()
            holder.type = StreamType.observable
            holder.sourceGenerator = observable
            holder.errorHandler = errorHandler
            return holder
        }

        fun <T> singleHolder(observable: (arguments: T?) -> SingleSource<*>, errorHandler: ((Throwable) -> Any?)? = null): ObservableSourceHolder<T> {
            val holder =  ObservableSourceHolder<T>()
            holder.type = StreamType.single
            holder.sourceGenerator = observable
            holder.errorHandler = errorHandler
            return holder
        }

        fun <T> completableHolder(observable: (arguments: T?) -> CompletableSource, errorHandler: ((Throwable) -> Any?)? = null): ObservableSourceHolder<T> {
            val holder =  ObservableSourceHolder<T>()
            holder.type = StreamType.completable
            holder.sourceGenerator = observable
            holder.errorHandler = errorHandler
            return holder
        }
    }

    fun getSourceAsObservable(args: T?): ObservableSource<*> {
        return sourceGenerator(args) as ObservableSource<*>
    }

    fun getSourceAsSingle(args: T?): SingleSource<*> {
        return sourceGenerator(args) as SingleSource<*>
    }

    fun getSourceAsCompletable(args: T?): CompletableSource {
        return sourceGenerator(args) as CompletableSource
    }
}